File: WePeiYang/app/src/main/java/com/twt/service/settings/SettingsActivity.java
package com.twt.service.settings;

import android.app.Activity;
import android.content.DialogInterface;
import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.preference.Preference;
import android.preference.PreferenceFragment;
import android.preference.SwitchPreference;
import android.support.annotation.Nullable;
import android.support.v7.app.AlertDialog;
import android.support.v7.app.AppCompatActivity;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.MenuItem;
import android.view.View;
import android.widget.EditText;
import android.widget.Toast;

import com.tencent.bugly.crashreport.CrashReport;
import com.twt.service.R;
import com.twt.service.home.HomeActivity;
import com.twt.service.update.UpdateManager;
import com.twt.wepeiyang.commons.cache.CacheProvider;
import com.twt.wepeiyang.commons.experimental.CommonPreferences;
import com.twt.wepeiyang.commons.network.RxErrorHandler;
import com.twtstudio.retrox.auth.api.AuthProvider;
import com.twtstudio.retrox.bike.service.BikeServiceProvider;
import com.twtstudio.retrox.tjulibrary.provider.TjuLibProvider;

import es.dmoral.toasty.Toasty;
import rx.android.schedulers.AndroidSchedulers;
import rx.schedulers.Schedulers;

/**
 * Created by retrox on 2017/2/21.
 */

public class SettingsActivity extends AppCompatActivity {

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_setting);

        setTitle("偏好设置");

        getSupportActionBar().setDisplayHomeAsUpEnabled(true);

//        CrashReport.testJavaCrash();

        SettingsFragment fragment = new SettingsFragment();
        fragment.setmContext(this);

        getFragmentManager().beginTransaction()
                .replace(R.id.settings_container, fragment)
                .commit();

    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if (item.getItemId() == android.R.id.home) {
            onBackPressed();
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    public void onBackPressed() {
        super.onBackPressed();
        Intent intent = new Intent(this, HomeActivity.class);
        intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
        startActivity(intent);
    }

    public static class SettingsFragment extends PreferenceFragment {

        private Activity mContext;

        public SettingsFragment() {
        }

        public void setmContext(Activity mContext) {
            this.mContext = mContext;
        }

        @Override
        public void onCreate(Bundle savedInstanceState) {
            super.onCreate(savedInstanceState);
            addPreferencesFromResource(R.xml.prefs);
            initPrefs();
        }

        @Override
        public void onResume() {
            super.onResume();
            Preference isBindLib = findPreference(getString(R.string.pref_is_bind_lib));
            isBindLib.setSummary(CommonPreferences.INSTANCE.isBindLibrary() ? "已绑定" : "未绑定");
        }

        public void initPrefs() {

            /**
             * 一键退学
             */

            Preference exitTjuPref = findPreference(getString(R.string.pref_is_exit_tju));
            int dropOutMode = CommonPreferences.INSTANCE.getDropOut();
            String dropOutSummary = "未操作";
            if (dropOutMode == 0) {
                dropOutSummary = "未操作";
            } else if (dropOutMode == 1) {
                dropOutSummary = "已退学";
            } else if (dropOutMode == 2) {
                dropOutSummary = "已复学";
            }
            exitTjuPref.setSummary("退学状态: " + dropOutSummary);

            /**
             * 这个代码有些冗杂 ... 因为一些内部类的调用关系没有理清楚但是先这样子吧
             */
            exitTjuPref.setOnPreferenceClickListener(new Preference.OnPreferenceClickListener() {

                @Override
                public boolean onPreferenceClick(Preference preference) {

                    if (CommonPreferences.INSTANCE.getDropOut() == 0 || CommonPreferences.INSTANCE.getDropOut() == 2) {
                        //退学
                        String[] items = {"我要打游戏！", "我要运动！", "我要睡觉！", "怎么样都好啦！"};

                        AlertDialog.Builder builder = new AlertDialog.Builder(mContext)
                                .setTitle("你为啥要退学呀？")
                                .setItems(items, new DialogInterface.OnClickListener() {
                                    @Override
                                    public void onClick(DialogInterface dialog, int which) {

                                        Toast.makeText(mContext, "正在办理...", Toast.LENGTH_SHORT).show();

                                        RealAuthApi.INSTANCE
                                                .dropOut(which + 1)
                                                .subscribeOn(Schedulers.io())
                                                .observeOn(AndroidSchedulers.mainThread())
                                                .subscribe(it -> {
                                                    if (!TextUtils.isEmpty(it.getData())) {
                                                        Toast.makeText(mContext, it.getData(), Toast.LENGTH_SHORT).show();

                                                        if (it.getData().equals("欢迎回来上学 (〃∀〃)")) {
                                                            CommonPreferences.INSTANCE.setDropOut(2);
                                                        } else if (it.getData().equals("退学成功 d(`･∀･)b")) {
                                                            CommonPreferences.INSTANCE.setDropOut(1);
                                                        }

                                                        //刷新下状态 --
                                                        int dropOutMode = CommonPreferences.INSTANCE.getDropOut();
                                                        String dropOutSummary = "未操作";
                                                        if (dropOutMode == 0) {
                                                            dropOutSummary = "未操作";
                                                        } else if (dropOutMode == 1) {
                                                            dropOutSummary = "已退学";
                                                        } else if (dropOutMode == 2) {
                                                            dropOutSummary = "已复学";
                                                        }
                                                        exitTjuPref.setSummary("退学状态: " + dropOutSummary);

                                                        //清空缓存 --- 课程表 GPA
                                                        CacheProvider.clearCache();

                                                        AuthProvider.INSTANCE.authSelf(false, true); //null做了处理的 刷新状态
                                                    }
                                                    dialog.dismiss();
                                                }, throwable -> {
                                                    AuthProvider.INSTANCE.authSelf(false, true); //null做了处理的
                                                    new RxErrorHandler().call(throwable);
                                                    dialog.dismiss();
                                                });
                                    }
                                });

                        builder.create().show();

                    } else {
                        AlertDialog.Builder dropInBuilder = new AlertDialog.Builder(mContext)
                                .setTitle("复学申请办理处")
                                .setMessage("浪子回头金不换啊...")
                                .setPositiveButton("我想好了...", new DialogInterface.OnClickListener() {
                                    @Override
                                    public void onClick(DialogInterface dialog, int which) {
                                        Toast.makeText(mContext, "正在办理...", Toast.LENGTH_SHORT).show();

                                        RealAuthApi.INSTANCE
                                                .dropOut(0)
                                                .subscribeOn(Schedulers.io())
                                                .observeOn(AndroidSchedulers.mainThread())
                                                .subscribe(it -> {
                                                    if (!TextUtils.isEmpty(it.getData())) {
                                                        Toast.makeText(mContext, it.getData(), Toast.LENGTH_SHORT).show();

                                                        if (it.getData().equals("欢迎回来上学 (〃∀〃)")) {
                                                            CommonPreferences.INSTANCE.setDropOut(2);
                                                        } else if (it.getData().equals("退学成功 d(`･∀･)b")) {
                                                            CommonPreferences.INSTANCE.setDropOut(1);
                                                        }

                                                        //刷新下状态 --
                                                        int dropOutMode = CommonPreferences.INSTANCE.getDropOut();
                                                        String dropOutSummary = "未操作";
                                                        if (dropOutMode == 0) {
                                                            dropOutSummary = "未操作";
                                                        } else if (dropOutMode == 1) {
                                                            dropOutSummary = "已退学";
                                                        } else if (dropOutMode == 2) {
                                                            dropOutSummary = "已复学";
                                                        }
                                                        exitTjuPref.setSummary("退学状态: " + dropOutSummary);

                                                        //清空缓存 --- 课程表 GPA
                                                        CacheProvider.clearCache();

                                                        AuthProvider.INSTANCE.authSelf(false, true);//null做了处理的 刷新状态
                                                    }
                                                    dialog.dismiss();
                                                }, throwable -> {
                                                    AuthProvider.INSTANCE.authSelf(false, true); //null做了处理的
                                                    new RxErrorHandler().call(throwable);
                                                    dialog.dismiss();
                                                });

                                    }
                                })
                                .setNegativeButton("我再浪会...", new DialogInterface.OnClickListener() {
                                    @Override
                                    public void onClick(DialogInterface dialog, int which) {
                                        dialog.dismiss();
                                    }
                                });

                        dropInBuilder.create().show();
                    }

                    return true;
                }
            });

            /**
             * 绑定模块
             */

            Preference libBindPref = findPreference(getString(R.string.pref_bind_settings));
            libBindPref.setOnPreferenceClickListener(new Preference.OnPreferenceClickListener() {
                @Override
                public boolean onPreferenceClick(Preference preference) {
                    AlertDialog.Builder builder = new AlertDialog.Builder(mContext)
                            .setMessage("是否要跳转到绑定页面")
                            .setPositiveButton("确定", new DialogInterface.OnClickListener() {
                                @Override
                                public void onClick(DialogInterface dialog, int which) {
                                    Intent intent = new Intent(mContext, BindActivity.class);
                                    mContext.startActivity(intent);
                                }
                            });
                    builder.create().show();
                    return true;
                }
            });


            Preference isBindTju = findPreference(getString(R.string.pref_is_bind_tju));
            isBindTju.setSummary(CommonPreferences.INSTANCE.isBindTju() ? "已绑定" : "未绑定");
            isBindTju.setOnPreferenceClickListener(new Preference.OnPreferenceClickListener() {
                @Override
                public boolean onPreferenceClick(Preference preference) {
                    AlertDialog.Builder builder = new AlertDialog.Builder(mContext)
                            .setTitle("办公网解绑")
                            .setMessage("是否要解绑办公网")
//                            .setCancelable(false)
                            .setPositiveButton("解绑", new DialogInterface.OnClickListener() {
                                @Override
                                public void onClick(DialogInterface dialog, int which) {
                                    if (CommonPreferences.INSTANCE.isBindTju()) {
                                        RealAuthApi.INSTANCE
                                                .unbindTju(CommonPreferences.INSTANCE.getTwtuname())
                                                .subscribeOn(Schedulers.io())
                                                .observeOn(AndroidSchedulers.mainThread())
                                                .doAfterTerminate(() -> AuthProvider.INSTANCE.authSelf(false, true))
                                                .subscribe(responseBody -> {
                                                    Toasty.success(mContext, "解绑成功！请重启微北洋", Toast.LENGTH_SHORT).show();
                                                }, new RxErrorHandler());
                                    } else {
                                        Toasty.warning(mContext, "你没绑定解绑啥？？？？？\n点击上面按钮进入绑定页面", Toast.LENGTH_SHORT).show();
                                    }
                                }
                            })
                            .setNegativeButton("再绑会...", new DialogInterface.OnClickListener() {
                                @Override
                                public void onClick(DialogInterface dialog, int which) {
                                    dialog.dismiss();
                                }
                            });
                    builder.create().show();
                    return false;
                }
            });

            Preference isBindLib = findPreference(getString(R.string.pref_is_bind_lib));
            isBindLib.setSummary(CommonPreferences.INSTANCE.isBindLibrary() ? "已绑定" : "未绑定");
            isBindLib.setOnPreferenceClickListener(new Preference.OnPreferenceClickListener() {
                @Override
                public boolean onPreferenceClick(Preference preference) {
                    AlertDialog.Builder builder = new AlertDialog.Builder(mContext)
                            .setTitle("图书馆解绑")
                            .setMessage("是否要解绑图书馆")
//                            .setCancelable(false)
                            .setPositiveButton("解绑", new DialogInterface.OnClickListener() {
                                @Override
                                public void onClick(DialogInterface dialog, int which) {
                                    if (CommonPreferences.INSTANCE.isBindLibrary()) {
                                        new TjuLibProvider(mContext).unbindLibrary(s -> {
                                            CommonPreferences.INSTANCE.setBindLibrary(false);
                                            Toasty.success(mContext, "解绑成功！请重启微北洋", Toast.LENGTH_SHORT).show();
                                            isBindLib.setSummary(CommonPreferences.INSTANCE.isBindLibrary() ? "已绑定" : "未绑定");
                                            dialog.dismiss();
                                        });
                                    } else {
                                        Toast.makeText(mContext, "你没绑定解绑啥？？？？？\n点击上面按钮进入绑定页面", Toast.LENGTH_SHORT).show();
                                    }

                                }
                            })
                            .setNegativeButton("再绑会...", new DialogInterface.OnClickListener() {
                                @Override
                                public void onClick(DialogInterface dialog, int which) {
                                    dialog.dismiss();
                                }
                            });
                    builder.create().show();
                    return false;
                }
            });

            Preference isBindBike = findPreference(getString(R.string.pref_is_bind_bike));
            isBindBike.setSummary(CommonPreferences.INSTANCE.isBindBike() ? "已绑定" : "未绑定");
            isBindBike.setOnPreferenceClickListener(new Preference.OnPreferenceClickListener() {
                @Override
                public boolean onPreferenceClick(Preference preference) {
                    AlertDialog.Builder builder = new AlertDialog.Builder(mContext)
                            .setTitle("自行车解绑")
                            .setMessage("是否要解绑自行车")
//                            .setCancelable(false)
                            .setPositiveButton("解绑", new DialogInterface.OnClickListener() {
                                @Override
                                public void onClick(DialogInterface dialog, int which) {
                                    if (CommonPreferences.INSTANCE.isBindBike()) {
                                        new BikeServiceProvider(mContext).unbind();
                                    } else {
                                        Toasty.warning(mContext, "你没绑定解绑啥？？？？？\n进入自行车模块完成绑定", Toast.LENGTH_SHORT).show();
                                    }
                                }
                            })
                            .setNegativeButton("再绑会...", new DialogInterface.OnClickListener() {
                                @Override
                                public void onClick(DialogInterface dialog, int which) {
                                    dialog.dismiss();
                                }
                            });
                    builder.create().show();
                    return false;
                }
            });

            Preference isDisplayBike = findPreference(getString(R.string.pref_is_display_bike));
            isDisplayBike.setOnPreferenceChangeListener(new Preference.OnPreferenceChangeListener() {
                @Override
                public boolean onPreferenceChange(Preference preference, Object newValue) {
                    if (newValue.equals(true)) {
                        Toast.makeText(mContext, "打开自行车模块以完成自行车功能的激活", Toast.LENGTH_SHORT).show();
                    }
                    return true;
                }
            });

//            Preference isChangeSourceEnabled = findPreference(getString(R.string.pref_is_switch_news_source));
//            if (BuildConfig.DEBUG){
//                isChangeSourceEnabled.setEnabled(true);
//            }else {
//                isChangeSourceEnabled.setEnabled(false);
//            }

            Preference devTalking = findPreference(getString(R.string.pref_dev_talking));
            devTalking.setOnPreferenceClickListener(new Preference.OnPreferenceClickListener() {
                @Override
                public boolean onPreferenceClick(Preference preference) {
                    Intent intent = new Intent(mContext, DevTalkActivity.class);
                    mContext.startActivity(intent);
                    return false;
                }
            });

            Preference feedback = findPreference(getString(R.string.pref_feedback));
            feedback.setOnPreferenceClickListener(new Preference.OnPreferenceClickListener() {
                @Override
                public boolean onPreferenceClick(Preference preference) {
                    Intent intent = new Intent(Intent.ACTION_VIEW);
                    intent.setData(Uri.parse("https://support.twtstudio.com/category/6/%E7%A7%BB%E5%8A%A8%E5%AE%A2%E6%88%B7%E7%AB%AF"));
//                    intent.createChooser(intent,"选择浏览器");
                    mContext.startActivity(intent);
                    return false;
                }
            });

            Preference contact = findPreference(getString(R.string.perf_contact_me));
            contact.setOnPreferenceClickListener(new Preference.OnPreferenceClickListener() {
                @Override
                public boolean onPreferenceClick(Preference preference) {
                    try {
                        String email = "mobile@twtstudio.com";
                        Uri uri = Uri.parse("mailto:" + email);
                        Intent emailIntent = new Intent(Intent.ACTION_SENDTO, uri);
                        mContext.startActivity(emailIntent);
                    } catch (Exception e) {
                        Toasty.error(mContext, "无法启动邮件发送APP", Toast.LENGTH_SHORT).show();
                        CrashReport.postCatchedException(e);
                        e.printStackTrace();
                    }
                    return false;
                }
            });

            Preference checkUpdate = findPreference(getString(R.string.pref_check_update));
            checkUpdate.setOnPreferenceClickListener(new Preference.OnPreferenceClickListener() {
                @Override
                public boolean onPreferenceClick(Preference preference) {
                    UpdateManager.getInstance().checkUpdate(getActivity());
                    return true;
                }
            });

            SwitchPreference isAutoCheckUpdate = (SwitchPreference) findPreference(getString(R.string.pref_is_auto_check_update));
            isAutoCheckUpdate.setChecked(UpdateManager.getInstance().getIsAutoCheck());
            isAutoCheckUpdate.setOnPreferenceChangeListener(new Preference.OnPreferenceChangeListener() {
                @Override
                public boolean onPreferenceChange(Preference preference, Object newValue) {
                    Boolean value = (Boolean) newValue;
                    UpdateManager.getInstance().setIsAutoCheck(value);
                    return true;
                }
            });

            Preference proxySetting = findPreference(getString(R.string.pref_proxy_settings));
            proxySetting.setOnPreferenceClickListener(new Preference.OnPreferenceClickListener() {
                @Override
                public boolean onPreferenceClick(Preference preference) {
                    LayoutInflater inflater = LayoutInflater.from(getActivity());
                    View dialogView = inflater.inflate(R.layout.dialog_proxy_setting,null,false);
                    final EditText proxyIP = dialogView.findViewById(R.id.edit_proxy_ip);
                    proxyIP.setText(CommonPreferences.INSTANCE.getProxyAddress());
//                    prxoyIP.setText();
                    final EditText proxyPort = dialogView.findViewById(R.id.edit_proxy_port);
                    proxyPort.setText(String.valueOf(CommonPreferences.INSTANCE.getProxyPort()));

                    AlertDialog dialog = new AlertDialog.Builder(getActivity()).setTitle("Proxy Settings")
                            .setView(dialogView)
                            .setPositiveButton("OK", new DialogInterface.OnClickListener() {
                                @Override
                                public void onClick(DialogInterface dialog, int which) {
                                    String ip = proxyIP.getText().toString();
                                    String port = proxyPort.getText().toString();
                                    CommonPreferences.INSTANCE.setProxyAddress(ip);
                                    CommonPreferences.INSTANCE.setProxyPort(Integer.parseInt(port));
                                    dialog.dismiss();
                                }
                            })
                            .setNegativeButton("Cancel", new DialogInterface.OnClickListener() {
                                @Override
                                public void onClick(DialogInterface dialog, int which) {
                                    dialog.dismiss();
                                }
                            })
                            .create();
                    dialog.show();
                    return true;
                }
            });


        }

        private void processExitTju() {
            // TODO: 23/03/2017 退学逻辑
        }
    }


}

File: WePeiYang/app/src/main/java/com/twt/service/settings/SettingsActivity.kt
package com.twt.service.settings

import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.preference.Preference
import android.preference.PreferenceFragment
import android.preference.SwitchPreference
import android.view.MenuItem
import android.widget.ImageView
import android.widget.Toast
import com.tencent.bugly.crashreport.CrashReport
import com.twt.service.R
import com.twt.service.update.UpdateManager
import es.dmoral.toasty.Toasty
import io.multimoon.colorful.CAppCompatActivity

/**
 * Created by retrox on 2017/2/21.
 */

class SettingsActivity : CAppCompatActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_setting)

        setSupportActionBar(findViewById(R.id.toolbar))
        title = ""
        findViewById<ImageView>(R.id.iv_activity_setting_back).setOnClickListener { onBackPressed() }
        fragmentManager.beginTransaction()
                .replace(R.id.settings_container, SettingsFragment())
                .commit()

    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == android.R.id.home) onBackPressed()
        return super.onOptionsItemSelected(item)
    }

    class SettingsFragment : PreferenceFragment() {

        override fun onCreate(savedInstanceState: Bundle?) {
            super.onCreate(savedInstanceState)
            addPreferencesFromResource(R.xml.prefs)
            initPrefs()
        }

        override fun onResume() {
            super.onResume()
//            val isBindLib = findPreference(getString(R.string.pref_is_bind_lib))
        }

        private fun initPrefs() {


            /**
             * 绑定模块
             */

//            val libBindPref = findPreference(getString(R.string.pref_bind_settings))
//            libBindPref.onPreferenceClickListener = Preference.OnPreferenceClickListener {
//                val builder = AlertDialog.Builder(activity)
//                        .setMessage("是否要跳转到绑定页面")
//                        .setPositiveButton("确定") { _, _ ->
//                            val intent = Intent(activity, BindActivity::class.java)
//                            activity.startActivity(intent)
//                        }
//                builder.create().show()
//                true
//            }
//
//            val isBindTju = findPreference(getString(R.string.pref_is_bind_tju))
//            isBindTju.setOnPreferenceClickListener {
//                val builder = AlertDialog.Builder(activity)
//                        .setTitle("办公网解绑")
//                        .setMessage("是否要解绑办公网")
//                        .setPositiveButton("解绑") { _, _ ->
//                            if (CommonPreferences.isBindTju) {
//                                RealBindAndDropOutService
//                                        .unbindTju(CommonPreferences.twtuname)
//                                        .subscribeOn(Schedulers.io())
//                                        .observeOn(AndroidSchedulers.mainThread())
//                                        .doAfterTerminate {
//                                            /* 由于后台的接口问题， 每次绑定解绑操作都要重新拿token，干脆用login接口做了假解绑，增加用户体验*/
//                                            login(CommonPreferences.twtuname, CommonPreferences.password) {
//                                                when (it) {
//                                                    is RefreshState.Success -> {
//                                                        authSelfLiveData.refresh(REMOTE)
//                                                    }
//                                                    is RefreshState.Failure -> {
//                                                        Toasty.error(activity, "发生错误 ${it.throwable.message}！${it.javaClass.name}").show()
//                                                    }
//                                                }
//                                            }
//                                        }
//                                        .subscribe(Action1 { Toasty.success(activity, "解绑成功！请重新绑定办公网", Toast.LENGTH_SHORT).show() }, RxErrorHandler())
//                            } else {
//                                Toasty.warning(activity, "你没绑定解绑啥？？？？？\n点击上面按钮进入绑定页面", Toast.LENGTH_SHORT).show()
//                            }
//                        }
//                        .setNegativeButton("再绑会...") { dialog, _ -> dialog.dismiss() }
//                builder.create().show()
//                false
//            }
//
//            val isBindLib = findPreference(getString(R.string.pref_is_bind_lib))
//            isBindLib.setOnPreferenceClickListener {
//                val builder = AlertDialog.Builder(activity)
//                        .setTitle("图书馆解绑")
//                        .setMessage("是否要解绑图书馆")
//                        .setPositiveButton("解绑") { dialog, _ ->
//                            if (CommonPreferences.isBindLibrary) {
//                                TjuLibProvider(activity).unbindLibrary {
//                                    login(CommonPreferences.twtuname, CommonPreferences.password) {
//                                        when (it) {
//                                            is RefreshState.Success -> {
//                                                authSelfLiveData.refresh(CacheIndicator.REMOTE)
//                                            }
//                                            is RefreshState.Failure -> {
//                                                Toasty.warning(activity, "你没绑定解绑啥？？？？？\n点击上面按钮进入绑定页面", Toast.LENGTH_SHORT).show()                                            }
//                                        }
//                                    }
//                                    Toasty.success(activity, "解绑成功！请重新绑定图书馆", Toast.LENGTH_SHORT).show()
//                                    dialog.dismiss()
//                                }
//                            } else {
//                                Toast.makeText(activity, "你没绑定解绑啥？？？？？\n点击上面按钮进入绑定页面", Toast.LENGTH_SHORT).show()
//                            }
//                        }
//                        .setNegativeButton("再绑会...") { dialog, _ -> dialog.dismiss() }
//                builder.create().show()
//                false
//            }
//
//            val isBindBike = findPreference(getString(R.string.pref_is_bind_bike))
////            isBindBike.summary = if (CommonPreferences.isBindBike) "已绑定" else "未绑定"
//            isBindBike.setOnPreferenceClickListener {
//                val builder = AlertDialog.Builder(activity)
//                        .setTitle("自行车解绑")
//                        .setMessage("是否要解绑自行车")
//                        .setPositiveButton("解绑") { _, _ ->
//                            if (CommonPreferences.isBindBike) {
//                                BikeServiceProvider(activity).unbind()
//                            } else {
//                                Toasty.warning(activity, "你没绑定解绑啥？？？？？\n进入自行车模块完成绑定", Toast.LENGTH_SHORT).show()
//                            }
//                        }
//                        .setNegativeButton("再绑会...") { dialog, _ -> dialog.dismiss() }
//                builder.create().show()
//                false
//            }
//
            val isDisplayBike = findPreference(getString(R.string.pref_is_display_bike))
            isDisplayBike.onPreferenceChangeListener = Preference.OnPreferenceChangeListener { _, newValue ->
                if (newValue == true) Toast.makeText(activity, "打开自行车模块以完成自行车功能的激活", Toast.LENGTH_SHORT).show()
                true
            }

            val devTalking = findPreference(getString(R.string.pref_dev_talking))
            devTalking.setOnPreferenceClickListener {
                val intent = Intent(activity, DevTalkActivity::class.java)
                activity.startActivity(intent)
                false
            }

            val feedback = findPreference(getString(R.string.pref_feedback))
            feedback.onPreferenceClickListener = Preference.OnPreferenceClickListener {
                val intent = Intent(Intent.ACTION_VIEW)
                intent.data = Uri.parse("https://support.twtstudio.com/category/6/%E7%A7%BB%E5%8A%A8%E5%AE%A2%E6%88%B7%E7%AB%AF")
                activity.startActivity(intent)
                false
            }

            val contact = findPreference(getString(R.string.pref_contact_me))
            contact.onPreferenceClickListener = Preference.OnPreferenceClickListener {
                try {
                    val email = "mobile@twtstudio.com"
                    val uri = Uri.parse("mailto:$email")
                    val emailIntent = Intent(Intent.ACTION_SENDTO, uri)
                    activity.startActivity(emailIntent)
                } catch (e: Exception) {
                    Toasty.error(activity, "无法启动邮件发送APP", Toast.LENGTH_SHORT).show()
                    CrashReport.postCatchedException(e)
                    e.printStackTrace()
                }

                false
            }

            val checkUpdate = findPreference(getString(R.string.pref_check_update))
            checkUpdate.setOnPreferenceClickListener {
                UpdateManager.getInstance().checkUpdate(activity)
                true
            }

            val isAutoCheckUpdate = findPreference(getString(R.string.pref_is_auto_check_update)) as SwitchPreference
            isAutoCheckUpdate.isChecked = UpdateManager.getInstance().isAutoCheck
            isAutoCheckUpdate.setOnPreferenceChangeListener { _, newValue ->
                UpdateManager.getInstance().isAutoCheck = newValue as Boolean
                true
            }

            val addGroup = findPreference(getString(R.string.pref_add_group))
            addGroup.setOnPreferenceClickListener {
                joinQQGroup("DqYP6KYECBOV36yk1_RBE6iVjXbKVv0R")
            }
        }

        private fun processExitTju() {
            // TODO: 23/03/2017 退学逻辑
        }

        /****************
         *
         * 发起添加群流程。群号：天外天用户社区(738068756) 的 key 为： DqYP6KYECBOV36yk1_RBE6iVjXbKVv0R
         * 调用 joinQQGroup(DqYP6KYECBOV36yk1_RBE6iVjXbKVv0R) 即可发起手Q客户端申请加群 天外天用户社区(738068756)
         *
         * @param key 由官网生成的key
         * @return 返回true表示呼起手Q成功，返回fals表示呼起失败
         */
        private fun joinQQGroup(key: String): Boolean {
            val intent = Intent()
            intent.data = Uri.parse("mqqopensdkapi://bizAgent/qm/qr?url=http%3A%2F%2Fqm.qq.com%2Fcgi-bin%2Fqm%2Fqr%3Ffrom%3Dapp%26p%3Dandroid%26k%3D$key")
            // 此Flag可根据具体产品需要自定义，如设置，则在加群界面按返回，返回手Q主界面，不设置，按返回会返回到呼起产品界面
            intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
            return try {
                startActivity(intent)
                true
            } catch (e: Exception) {
                // 未安装手Q或安装的版本不支持
                Toasty.error(activity, "未安装手Q或安装的版本不支持", Toast.LENGTH_SHORT).show()
                false
            }

        }

    }


}
